monitoring/rpc_stats_interceptor.go

```diff
@@ -26,89 +25,81 @@ import (
+	ReqSuccessLatency Histogram
+	ReqErrorCount     Counter
+	ReqErrorLatency   Histogram
 }
 
 // NewRPCStatsInterceptor creates a new RPCStatsInterceptor for the given application/component, with
 // a specified time source.
-func NewRPCStatsInterceptor(timeSource util.TimeSource, application, component string) *RPCStatsInterceptor {
-	return &RPCStatsInterceptor{baseName: fmt.Sprintf("%s/%s", application, component), timeSource: timeSource,
-		handlerRequestCountMap:            new(expvar.Map).Init(),
-		handlerRequestSucceededCountMap:   new(expvar.Map).Init(),
-		handlerRequestErrorCountMap:       new(expvar.Map).Init(),
-		handlerRequestSucceededLatencyMap: new(expvar.Map).Init(),
-		handlerRequestFailedLatencyMap:    new(expvar.Map).Init()}
-}
-
-func (r RPCStatsInterceptor) nameForMap(name string) string {
-	return fmt.Sprintf("%s/%s", r.baseName, name)
+func NewRPCStatsInterceptor(timeSource util.TimeSource, prefix string, mf MetricFactory) *RPCStatsInterceptor {
+	interceptor := RPCStatsInterceptor{
+		prefix:            prefix,
+		timeSource:        timeSource,
+		ReqCount:          mf.NewCounter(prefixedName(prefix, reqCountName), "Number of requests", []string{methodName}),
+		ReqSuccessCount:   mf.NewCounter(prefixedName(prefix, reqSuccessCountName), "Number of successful requests", []string{methodName}),
+		ReqSuccessLatency: mf.NewHistogram(prefixedName(prefix, reqSuccessLatencyName), "Latency of successful requests", []string{methodName}),
+		ReqErrorCount:     mf.NewCounter(prefixedName(prefix, reqErrorCountName), "Number of errored requests", []string{methodName}),
+		ReqErrorLatency:   mf.NewHistogram(prefixedName(prefix, reqErrorLatencyName), "Latency of errored requests", []string{methodName}),
+	}
+	return &interceptor
 }
 
-// Publish must be called for stats to be visible. The expvar framework will prevent
-// multiple calls to Publish from succeeding.
-func (r RPCStatsInterceptor) Publish() {
-	expvar.Publish(r.nameForMap(requestCountMapName), r.handlerRequestCountMap)
-	expvar.Publish(r.nameForMap(requestSucceededCountMapName), r.handlerRequestSucceededCountMap)
-	expvar.Publish(r.nameForMap(requestErrorCountMapName), r.handlerRequestErrorCountMap)
-	expvar.Publish(r.nameForMap(requestSucceededLatencyMapName), r.handlerRequestSucceededLatencyMap)
-	expvar.Publish(r.nameForMap(requestFailedLatencyMapName), r.handlerRequestFailedLatencyMap)
+func prefixedName(prefix, name string) string {
+	return fmt.Sprintf("%s_%s", prefix, name)
 }
 
-func (r RPCStatsInterceptor) recordFailureLatency(method string, startTime time.Time) {
+func (r *RPCStatsInterceptor) recordFailureLatency(labels []string, startTime time.Time) {
 	latency := r.timeSource.Now().Sub(startTime)
-	r.handlerRequestErrorCountMap.Add(method, 1)
-	r.handlerRequestFailedLatencyMap.Add(method, latency.Nanoseconds()/nanosToMillisDivisor)
+	r.ReqErrorCount.Inc(labels...)
+	r.ReqErrorLatency.Observe(float64(latency.Nanoseconds()/int64(time.Millisecond)), labels...)
```

Comment:
`latency.Nanoseconds()` is an `int64`, but you're right, it can be simplified to `float64(latency/time.Millisecond)` - done.